File: 20-valid-parentheses/20-valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        if (s == null || s.length() == 0) {
            return true;
        }
        HashMap<Character, Character> map = new HashMap<>();
        map.put('(', ')');
        map.put('{', '}');
        map.put('[', ']');
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char top = stack.pop();
                if (map.get(top) != c) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/20-valid-parentheses.py
"""
Create a charMap and a stack
add the key value to stack if str[i] in key i.e. if str[i] == "{", then add "}" to stack
else, check if the top of stack is equal to str[i] i.e. if str[i] == "}", check if the top of the stack is "}". If it's not, return false
i.e. {,[,(,),],},}
"""
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = []
        charMap = {"{": "}", "(": ")", "[": "]"}
        
        for c in s:
            if c in charMap:
                stack.append(charMap[c])
            elif stack:
                p = stack.pop() 
                if c != p:
                    return False
            else:
                return False
                    
        
        if len(stack) > 0: 
            return False
        
        return True